node/src/blockchain/errors/validation_status.rs

```diff
@@ -151,4 +180,67 @@ mod tests {
             .get(&BlockchainErrorKind::AppRpc(AppRpcErrorKind::Signing));
         assert_eq!(other_error_stats, None);
     }
+
+    #[test]
+    fn previous_attempts_hash_works_correctly() {
+        let now = SystemTime::now();
+        let clock = ValidationFailureClockMock::default()
+            .now_result(now)
+            .now_result(now)
+            .now_result(now + Duration::from_secs(2));
+        let attempts1 = PreviousAttempts::new(
+            BlockchainErrorKind::AppRpc(AppRpcErrorKind::Decoder),
+            &clock,
+        );
+        let attempts2 = PreviousAttempts::new(
+            BlockchainErrorKind::AppRpc(AppRpcErrorKind::Decoder),
+            &clock,
+        );
+        let attempts3 =
+            PreviousAttempts::new(BlockchainErrorKind::AppRpc(AppRpcErrorKind::Io), &clock);
+        let hash1 = {
+            let mut hasher = DefaultHasher::new();
+            attempts1.hash(&mut hasher);
+            hasher.finish()
+        };
+        let hash2 = {
+            let mut hasher = DefaultHasher::new();
+            attempts2.hash(&mut hasher);
+            hasher.finish()
+        };
+        let hash3 = {
+            let mut hasher = DefaultHasher::new();
+            attempts3.hash(&mut hasher);
+            hasher.finish()
+        };
+
+        assert_eq!(hash1, hash2);
+        assert_ne!(hash1, hash3);
+    }
+
+    #[test]
+    fn previous_attempts_ordering_works_correctly_with_mock() {
+        let now = SystemTime::now();
+        let clock = ValidationFailureClockMock::default()
+            .now_result(now)
+            .now_result(now + Duration::from_secs(1))
+            .now_result(now + Duration::from_secs(2))
+            .now_result(now + Duration::from_secs(3));
+        let mut attempts1 = PreviousAttempts::new(
+            BlockchainErrorKind::AppRpc(AppRpcErrorKind::Decoder),
+            &clock,
+        );
+        let mut attempts2 =
+            PreviousAttempts::new(BlockchainErrorKind::AppRpc(AppRpcErrorKind::Io), &clock);
+        attempts1 = attempts1.add_attempt(
```

Comment:
Why can't you do this together with the beginning of the setup of attempts1? Do not interrupt it with attempts2. 
If you wanted to shuffle with the order of the timestamps, just prepare them differently ordered at the mock.